### PR TITLE
[MSFT] Wire cleanup pass

### PR DESCRIPTION
**Pull request by teqdruid**
Another step towards #2365. This pass 'bubbles up' wires which are merely pass throughs in a given module. It then 'sinks down' wires which are looped back in the instantiation. Together with the entity movement piece, this effectively moves the wires as well. Does not handle wire manipulation operations, which also need to be moved/copied.

### CODE REVIEW

**Comment by teqdruid** [comment]
Refactoring PRs on which this PR depends (includes): #2415 and #2416.

File: test/Dialect/MSFT/partition.mlir [hunk 1]
```diff
@@ -1,39 +1,57 @@
 // RUN: circt-opt %s --msft-partition -verify-diagnostics -split-input-file | FileCheck %s
+// RUN: circt-opt %s --msft-partition --msft-wire-cleanup -verify-diagnostics -split-input-file | FileCheck --check-prefix=CLEANUP %s
 
-msft.module @top {} (%clk : i1) -> () {
+msft.module @top {} (%clk : i1) -> (out1: i2, out2: i2) {
```
Comment (teqdruid) [inline test/Dialect/MSFT/partition.mlir:4]: I moved to `i2` (from `i1`) to differentiate data from clock by the type. Catch more bugs that way.

**Review by mikeurbach — APPROVED** [review]
Some minor things, but looks good to me.

File: lib/Dialect/MSFT/MSFTPasses.cpp [hunk 2]
```diff
@@ -722,6 +722,166 @@ std::unique_ptr<Pass> createPartitionPass() {
...
+      assert(inputPortNum <= oldInst.getNumOperands());
+      oldInst.getResult(outputPortNum)
+          .replaceAllUsesWith(oldInst.getOperand(inputPortNum));
+    }
+    // Use a sort-merge-join approach to figure out the operand mapping on the
+    // fly.
+    size_t mergeCtr = 0;
+    for (size_t operNum = 0, e = oldInst.getNumOperands(); operNum < e;
+         ++operNum) {
+      if (operNum < inputPortsToRemove.size() &&
```
[the remark this answers is not in the document]
Comment (teqdruid) [reply lib/Dialect/MSFT/MSFTPasses.cpp:809]: These are the sorts of things which happen when you're working in 20 minute increments...

**Comment by teqdruid** [comment]
@mikeurbach thanks for the review!

**Pull request merged**